### PR TITLE
14 add renoir dataset

### PR DESCRIPTION
**Pull request by FedericoCalandra**
Ho aggiunto la classe RENOIRDatasetBuilder. Costruisce il test set per valutare le performance dei modelli complessi che andremo ad implementare. 

Per farlo funzionare è necessario:
1) scaricare il dataset da questo link (va bene uno tra Canon T3i e Xiaomi Mi3 o entrambi, versione aligned)
2) decomprimere la cartella
3) il percorso della cartella va passato al costruttore della classe: `__init__(self, dataset_folder, dataset_type)`
4) dataset_type può essere 'Mi3_Aligned' o 'T3i_Aligned'

Nota: il dataset è molto grande, non sono sicuro di quale sia il modo migliore per importarlo (su GitHub?)

### CODE REVIEW

**Comment by peiva-git** [comment]
Su GitHub non penso sia fattibile. Un modo sarebbe caricarlo su [Kaggle](https://www.kaggle.com/) e usare dopo le loro API REST per recuperarlo. Non ho mai provato ma credo non sia troppo difficile.
Altrimenti, più semplicemente, lasciamo che bisogna scaricarlo manualmente.

**Comment by peiva-git** [comment]
Temo però che leggere tutte le immagini del dataset (ovvero caricarle in memoria) al momento della creazione dell'istanza della classe non funzionerà, perché il dataset è troppo grande. Bisognerà utilizzare la classe [Sequence](https://www.tensorflow.org/api_docs/python/tf/keras/utils/Sequence) di Keras.

**Review by peiva-git — CHANGES_REQUESTED** [review]
Vedi commento in #40

**Comment by FedericoCalandra** [comment]
Ho fatto un po' di prove e funziona bene ma il caricamento in memoria è effettivamente piuttosto lento (non meno di un paio di minuti) e satura completamente i miei poveri 8 GB di ram. Seguo il consiglio e implemento una nuova versione di RENOIRDatasetBuilder usando la classe Sequence di Keras.

**Comment by FedericoCalandra** [comment]
Ho aggiunto la nuova versione. Purtroppo non ho notato significative differenze nei tempi di esecuzione ma forse è dovuto al fatto che semplicemente ho un PC vecchio

File: dlproject/data/renoir_dataset.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            raise ValueError("Invalid dataset type. Choose 'Mi3_Aligned' or 'T3i_Aligned'.")
+
+        self.dataset_folder = dataset_folder
+        self.dataset_type = dataset_type
+        self.batch_size = batch_size
+        self.target_size = target_size
+        self.test_data_paths = self.__get_test_data_paths()
+
+    def __len__(self):
+        return math.ceil(len(self.test_data_paths) / self.batch_size)
```
Comment (peiva-git) [inline dlproject/data/renoir_dataset.py:21]: Non è necessario usare np.ceil

File: dlproject/data/renoir_dataset.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    def __getitem__(self, index):
+        low = index * self.batch_size
+        high = min(low + self.batch_size, len(self.test_data_paths))
```
Comment (peiva-git) [inline dlproject/data/renoir_dataset.py:25]: Nel caso in cui l'ultimo batch non abbia dimensione `batch_size`. Non necessariamente il numero di immagini è divisibile per `batch_size`.

File: requirements.txt [hunk 3]
```diff
@@ -1,3 +1,5 @@
 tensorflow==2.14.*
 numpy==1.26.*
 matplotlib==3.8.*
+setuptools~=60.2.0
```
Comment (peiva-git) [inline requirements.txt:4]: Perché non l'ultima versione 68.2.2?

**Review by peiva-git — CHANGES_REQUESTED** [review]
Ben fatto, ho applicato alcune correzioni, ho solo un dubbio prima di approvare, vedi commenti.

File: requirements.txt (hunk 3 above)
Comment (FedericoCalandra) [reply requirements.txt]: Risolto

**Review by peiva-git — APPROVED** [review]
👍
